Add Gymnasium support

araffin opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->

Like https://github.com/DLR-RM/stable-baselines3/pull/780 but with gymnasium as default and compat wrapper on (would close https://github.com/DLR-RM/stable-baselines3/issues/871 too).

Installation:
```
pip install git+https://github.com/DLR-RM/stable-baselines3@feat/gymnasium-support
pip install git+https://github.com/Stable-Baselines-Team/stable-baselines3-contrib@feat/gymnasium-support
```

Documentation is available here: https://stable-baselines3.readthedocs.io/en/feat-gymnasium-support/

There is an **alpha version on pypi**:
```
# sb3 contrib installs sb3 too
pip install "sb3_contrib>=2.0.0a1" --upgrade
pip install "stable_baselines3>=2.0.0a1" --upgrade
```


Some new issues:
- typing annotations for wrappers is broken (since https://github.com/Farama-Foundation/Gymnasium/pull/39), not clear what should be the new annotations as gymnasium doesn't even do it for its own wrappers, should be fixed by https://github.com/Farama-Foundation/Gymnasium/pull/337
- warnings when importing gymnasium, should be fixed by shimmy 0.2.1
- slower import (see https://github.com/Farama-Foundation/Gymnasium/issues/314), should be fixed in next Gymnasium release
- shimmy doesn't allow gym 0.21 but should (https://github.com/Farama-Foundation/Shimmy/pull/29), should be fixed in shimmy 0.2.1
- test fails when checking for discrete env because of https://github.com/Farama-Foundation/Gymnasium/pull/141
- spec behavior is broken (and would break RL Zoo too), see https://github.com/DLR-RM/stable-baselines3/pull/1327#issuecomment-1445219678

Missing:
- [x]  update changelog
- [x] update notebooks (rename imports as they should already work with this version)
- [x] update tutorials
- [x] Cleanup GymStep26Return


Related but due to pygame: https://github.com/pygame/pygame/issues/3572

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
<!--- You can use the syntax `closes #100` if this solves the issue #100 -->
- [ ] I have raised an issue to propose this change ([required](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) for new features and bug fixes)

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)

closes #840 #871 
closes #271 
closes #1156 
closes #1324 
closes https://github.com/DLR-RM/stable-baselines3/issues/1406
closes https://github.com/DLR-RM/stable-baselines3/issues/1315

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] I've read the [CONTRIBUTION](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) guide (**required**)
- [ ] I have updated the changelog accordingly (**required**).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.
- [ ] I have opened an associated PR on the [SB3-Contrib repository](https://github.com/Stable-Baselines-Team/stable-baselines3-contrib) (if necessary)
- [ ] I have opened an associated PR on the [RL-Zoo3 repository](https://github.com/DLR-RM/rl-baselines3-zoo) (if necessary)
- [ ] I have reformatted the code using `make format` (**required**)
- [ ] I have checked the codestyle using `make check-codestyle` and `make lint` (**required**)
- [ ] I have ensured `make pytest` and `make type` both pass. (**required**)
- [ ] I have checked that the documentation builds using `make doc` (**required**)

Note: You can run most of the checks using `make commit-checks`.

Note: we are using a maximum length of 127 characters per line

<!--- This Template is an edited version of the one from https://github.com/evilsocket/pwnagotchi/ -->

pseudo-rnd-thoughts commented:
@araffin Could you test that this PR (https://github.com/Farama-Foundation/Gymnasium/pull/337) fixes the issues you list above (except the `Discrete` issue)

araffin commented:
> @araffin Could you test that this PR ([Farama-Foundation/Gymnasium#337](https://github.com/Farama-Foundation/Gymnasium/pull/337)) fixes the issues you list above (except the `Discrete` issue)

thanks =)
tested in https://github.com/DLR-RM/stable-baselines3/pull/1327/commits/ad48559b4e704f89896edb91bc3621e9236d267c

araffin commented:
Error on the CI is due to https://github.com/pygame/pygame/issues/3572 (can be fixed by downgrading to pygame 2.1.2)

araffin commented on docs/conda_env.yml:
```diff
@@ -4,11 +4,11 @@ channels:
+  - pytorch=1.11.0=py3.7_cpu_0
   - pip:
-    - gym==0.21
+    - gym==0.26
```
TODO(antonin): add gymnasium and shimmy

pseudo-rnd-thoughts commented:
> Error on the CI is due to https://github.com/pygame/pygame/issues/3572 (can be fixed by downgrading to pygame 2.1.2)

If we changed over to pygame ce 2.1.3 does this fix the issue? If so, we might change such that we can stop using a dev version

araffin commented:
> > Error on the CI is due to [pygame/pygame#3572](https://github.com/pygame/pygame/issues/3572) (can be fixed by downgrading to pygame 2.1.2)
> 
> If we changed over to pygame ce 2.1.3 does this fix the issue? If so, we might change such that we can stop using a dev version

no, the comes actually comes from 2.1.3 (see logs on github), it has to be fixed on the pygame side (I updated the requirements in setup to download the correct version for python 3.7 in https://github.com/DLR-RM/stable-baselines3/pull/1327/commits/75c22667c40c9b4285ca044913b37dc1a6253ad8).

pseudo-rnd-thoughts commented:
Ok, could you link to a comment on 
> test fails when checking for discrete env because of https://github.com/Farama-Foundation/Gymnasium/pull/141

araffin commented:
> Ok, could you link to a comment on

test was fixed in https://github.com/DLR-RM/stable-baselines3/pull/1327/commits/bfcb0ee2c4f2c3b055ac1074e685a471e1d4f45d (probably part of the small breaking changes not mentioned in Gymnasium doc)

pseudo-rnd-thoughts commented:
Apologies, the v0.27 release notes are updated to make this more clear

pseudo-rnd-thoughts commented:
@araffin Would you be able to test with PR https://github.com/Farama-Foundation/Gymnasium/pull/292
We aim for this to be the last PR before we release v0.28 and just want to check that no new issue are introduced by it.
The PR changes the implementation of `make` when a `EnvSpec` is used rather than a string as normal.

araffin commented:
> @araffin Would you be able to test with PR [Farama-Foundation/Gymnasium#292](https://github.com/Farama-Foundation/Gymnasium/pull/292) We aim for this to be the last PR before we release v0.28 and just want to check that no new issue are introduced by it. The PR changes the implementation of `make` when a `EnvSpec` is used rather than a string as normal.

that the case in the latest commit (we point to master)

araffin commented:
@pseudo-rnd-thoughts the current tests fails with Gymnasium master as the `env.spec` cannot be overridden and there is apparently no simple way to remove the time limit (passing `max_episode_steps=None` to `make()` doesn't remove it, nor `max_episode_steps=-1`)

pseudo-rnd-thoughts commented:
> @pseudo-rnd-thoughts the current tests fails with Gymnasium master as the `env.spec` cannot be overridden and there is apparently no simple way to remove the time limit (passing `max_episode_steps=None` to `make()` doesn't remove it, nor `max_episode_steps=-1`)

If you pass an environment spec then none of the make keywords will have an effect. If you want the old implementation of make with an environment spec, you now need to do `make(env.spec.id)`

araffin commented:
> If you pass an environment spec then none of the make keywords will have an effect. If you want the old implementation of make with an environment spec, you now need to do make(env.spec.id).

It seems that the spec is broken now, this breaks also the RL Zoo :disappointed: 

Before (gym 0.26 / gymnasium 0.27):
```python
import gym

spec = gym.spec("CartPole-v1")

# <TimeLimit<CartPoleEnv<CartPole-v1>>>
spec.make()

spec.max_episode_steps = None

# <OrderEnforcing<CartPoleEnv<CartPole-v1>>>
spec.make()
```

after:
```python
import gymnasium as gym

spec = gym.spec("CartPole-v1")

# <gymnasium.envs.classic_control.cartpole.CartPoleEnv at ...>
spec.make()

spec.max_episode_steps = None

# <gymnasium.envs.classic_control.cartpole.CartPoleEnv at ...>
spec.make()

# but
# <TimeLimit<OrderEnforcing<PassiveEnvChecker<CartPoleEnv<CartPole-v1>>>>>
gym.make("CartPole-v1")

```

pseudo-rnd-thoughts commented:
Damn it, we considered a number of backward compatible problems but didn't know that one. We will not make a release until all of the backward compatible issues are resolved to the best we can or take a different approach.

elliottower commented:
Appreciate all the work you've done, but one additional bug I've found with this current feature branch is that the stable_baselines3.common vec_env methods still expect `dones`to be returned from`step()`, whereas gymnasium has replaced this with `terminations` and `truncations`. The reset() function and most other things worked though.

araffin commented:
@elliottower Could you open an issue following the template and provide minimal code to reproduce the error + traceback?
The VecEnv api is on purpose different from the gym api (we explain it in the doc of that pr).

pseudo-rnd-thoughts commented:
@araffin I have looked into the backward compatibility issue. 
We have this proposal to fix the issue - https://github.com/Farama-Foundation/Gymnasium/pull/355
 
This is not perfectly backward compatible with issues if users do `env.unwrapped.spec` which I believe is very rare but would be interested in this raises any new issues? 
We do have an alternative solution that we are comparing to this one, I just wanted to see if this one works.

AlejandroCN7 commented:
Hi @araffin,
Will this PR be merged in master?

araffin commented:
> Hi @araffin, Will this PR be merged in master?

See https://github.com/DLR-RM/stable-baselines3/pull/780#issuecomment-1442498721

In the meantime, if you need that branch, you can already use it (see instructions at the top).

Kallinteris-Andreas commented:
@araffin, does the current version properly handle `terminated` and `truncated` steps?

For example, I am running this:
```
import gymnasium as gym

from stable_baselines3 import TD3

env = gym.make("Hopper-v4")

model = TD3("MlpPolicy", env, verbose=1)
model.learn(total_timesteps=10_000)

vec_env = model.get_env()
obs = vec_env.reset()
for i in range(1000):
    action, _states = model.predict(obs, deterministic=True)
    obs, reward, done, info = vec_env.step(action)
```

and `vec_env.step(action)` returns `done` and not `terminated/truncated`

Thanks!!

araffin commented:
> @araffin, does the current version properly handle terminated and truncated steps?

yes https://stable-baselines3.readthedocs.io/en/feat-gymnasium-support/guide/vec_envs.html#vecenv-api-vs-gym-api

and SB3 already handles truncation/timeouts properly with gym 0.21 (see https://github.com/DLR-RM/stable-baselines3/issues/633 and https://github.com/DLR-RM/stable-baselines3/issues/284)

jwarley commented:
Thank you for adding this!

I'm trying to use the alpha version with minigrid and running into some issues:

```
import minigrid
import gymnasium as gym
from stable_baselines3 import PPO
from stable_baselines3.ppo import MlpPolicy as PPOMlp

env = gym.make('MiniGrid-DistShift1-v0')
agent = PPO(PPOMlp, env, verbose=1)
```

results in the following:

```
Using cpu device
Wrapping the env with a `Monitor` wrapper
Wrapping the env in a DummyVecEnv.
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[11], line 7
      4 from stable_baselines3.ppo import MlpPolicy as PPOMlp
      6 env = gym.make('MiniGrid-DistShift1-v0', render_mode="rgb_array")
----> 7 agent = PPO(PPOMlp, env, verbose=1)

File ~/opt/anaconda3/envs/sb3test/lib/python3.9/site-packages/stable_baselines3/ppo/ppo.py:101, in PPO.__init__(self, policy, env, learning_rate, n_steps, batch_size, n_epochs, gamma, gae_lambda, clip_range, clip_range_vf, normalize_advantage, ent_coef, vf_coef, max_grad_norm, use_sde, sde_sample_freq, target_kl, tensorboard_log, policy_kwargs, verbose, seed, device, _init_setup_model)
     75 def __init__(
     76     self,
     77     policy: Union[str, Type[ActorCriticPolicy]],
   (...)
     99     _init_setup_model: bool = True,
    100 ):
--> 101     super().__init__(
    102         policy,
    103         env,
    104         learning_rate=learning_rate,
    105         n_steps=n_steps,
    106         gamma=gamma,
    107         gae_lambda=gae_lambda,
    108         ent_coef=ent_coef,
    109         vf_coef=vf_coef,
    110         max_grad_norm=max_grad_norm,
    111         use_sde=use_sde,
    112         sde_sample_freq=sde_sample_freq,
    113         tensorboard_log=tensorboard_log,
    114         policy_kwargs=policy_kwargs,
    115         verbose=verbose,
    116         device=device,
    117         seed=seed,
    118         _init_setup_model=False,
    119         supported_action_spaces=(
    120             spaces.Box,
    121             spaces.Discrete,
    122             spaces.MultiDiscrete,
    123             spaces.MultiBinary,
    124         ),
    125     )
    127     # Sanity check, otherwise it will lead to noisy gradient and NaN
    128     # because of the advantage normalization
    129     if normalize_advantage:

File ~/opt/anaconda3/envs/sb3test/lib/python3.9/site-packages/stable_baselines3/common/on_policy_algorithm.py:75, in OnPolicyAlgorithm.__init__(self, policy, env, learning_rate, n_steps, gamma, gae_lambda, ent_coef, vf_coef, max_grad_norm, use_sde, sde_sample_freq, tensorboard_log, monitor_wrapper, policy_kwargs, verbose, seed, device, _init_setup_model, supported_action_spaces)
     53 def __init__(
     54     self,
     55     policy: Union[str, Type[ActorCriticPolicy]],
   (...)
     73     supported_action_spaces: Optional[Tuple[spaces.Space, ...]] = None,
     74 ):
---> 75     super().__init__(
     76         policy=policy,
     77         env=env,
     78         learning_rate=learning_rate,
     79         policy_kwargs=policy_kwargs,
     80         verbose=verbose,
     81         device=device,
     82         use_sde=use_sde,
     83         sde_sample_freq=sde_sample_freq,
     84         support_multi_env=True,
     85         seed=seed,
     86         tensorboard_log=tensorboard_log,
     87         supported_action_spaces=supported_action_spaces,
     88     )
     90     self.n_steps = n_steps
     91     self.gamma = gamma

File ~/opt/anaconda3/envs/sb3test/lib/python3.9/site-packages/stable_baselines3/common/base_class.py:165, in BaseAlgorithm.__init__(self, policy, env, learning_rate, policy_kwargs, tensorboard_log, verbose, device, support_multi_env, monitor_wrapper, seed, use_sde, sde_sample_freq, supported_action_spaces)
    163 if env is not None:
    164     env = maybe_make_env(env, self.verbose)
--> 165     env = self._wrap_env(env, self.verbose, monitor_wrapper)
    167     self.observation_space = env.observation_space
    168     self.action_space = env.action_space

File ~/opt/anaconda3/envs/sb3test/lib/python3.9/site-packages/stable_baselines3/common/base_class.py:216, in BaseAlgorithm._wrap_env(env, verbose, monitor_wrapper)
    214     if verbose >= 1:
    215         print("Wrapping the env in a DummyVecEnv.")
--> 216     env = DummyVecEnv([lambda: env])
    218 # Make sure that dict-spaces are not nested (not supported)
    219 check_for_nested_spaces(env.observation_space)

File ~/opt/anaconda3/envs/sb3test/lib/python3.9/site-packages/stable_baselines3/common/vec_env/dummy_vec_env.py:44, in DummyVecEnv.__init__(self, env_fns)
     41 obs_space = env.observation_space
     42 self.keys, shapes, dtypes = obs_space_info(obs_space)
---> 44 self.buf_obs = OrderedDict([(k, np.zeros((self.num_envs, *tuple(shapes[k])), dtype=dtypes[k])) for k in self.keys])
     45 self.buf_dones = np.zeros((self.num_envs,), dtype=bool)
     46 self.buf_rews = np.zeros((self.num_envs,), dtype=np.float32)

File ~/opt/anaconda3/envs/sb3test/lib/python3.9/site-packages/stable_baselines3/common/vec_env/dummy_vec_env.py:44, in <listcomp>(.0)
     41 obs_space = env.observation_space
     42 self.keys, shapes, dtypes = obs_space_info(obs_space)
---> 44 self.buf_obs = OrderedDict([(k, np.zeros((self.num_envs, *tuple(shapes[k])), dtype=dtypes[k])) for k in self.keys])
     45 self.buf_dones = np.zeros((self.num_envs,), dtype=bool)
     46 self.buf_rews = np.zeros((self.num_envs,), dtype=np.float32)

TypeError: 'NoneType' object is not iterable
```

araffin commented:
> I'm trying to use the alpha version with minigrid and running into some issues:

Please open a separate issue and make sure to use the env checker first.

jkterry1 commented:
Just so I know- what's the plan for making a release that includes this PR?

araffin commented:
> Just so I know- what's the plan for making a release that includes this PR?

Alpha release is already on pypi since 15 days, stable release will be in some weeks if everything is fine.
